https by default

lool opened the pull request:
This avoids using https where we can and hardcoding the default mirror for debootstrap.
- **refactor(debootstrap_action): No default mirror**
- **refactor(debootstrap_action): Drop APT sources gen**
- **refactor(debootstrap_action): Add unit tests**
- **tests: Add integration tests for debootstrap**
- **fix: Use https links where possible**
- **refactor: Clarify default mirror for mmdebstrap**

The only remaining file with http:// URLs is the Apache 2.0 license, but it's probably a bad idea to patch it for compliance tooling to work.

lool committed:
refactor: Clarify default mirror for mmdebstrap

debos doesn't actually set a default if the user doesn't provide one;
clarify that the current default mirror for mmdebstrap is defined in
that tool and avoid referring to the current effective default as it
might change in mmdebstrap.

Signed-off-by: Loïc Minier <loic.minier@oss.qualcomm.com>

lool committed:
fix: Use https links where possible

Use secure https:// links instead of http:// links wherever possible.
Update bmaptool URL to a working one.

Signed-off-by: Loïc Minier <loic.minier@oss.qualcomm.com>

lool commented:
For mmdebstrap, we could actually override its defaults to have debos guarantee https sources.list in its default configuration – let me know if you want me to send that change.

lool commented:
I realize this PR might seem petty, I'm sending this because of the result of an actual cybersecurity assessment of an image based of qcom-deb-images :)

sjoerdsimons commented:
The suggested url by deb.debian.org is http, not https ; I'd be curious to know what the assesments is, in the end the signing/validation is done by the gpg checks tracking back to the packages file, not by the network transport.

when building for non-native architectures using https can also cause image builds to be quite a bit slower as the whole encryption is emulated

sjoerdsimons commented:
> > The suggested url by deb.debian.org is http, not https ; I'd be curious to know what the assesments is, in the end the signing/validation is done by the gpg checks tracking back to the packages file, not by the network transport.
> 
> When you say the suggested url, you reference the text file hosted on deb.debian.org itself? 

Yeah; Also:

* It's the default for debootstrap: https://salsa.debian.org/installer-team/debootstrap/-/blob/master/debootstrap?ref_type=heads#L53
* and the default for mmdebstrap: https://gitlab.mister-muffin.de/josch/mmdebstrap/src/branch/main/mmdebstrap#L5319 
* and the reference manual: https://www.debian.org/doc/manuals/debian-reference/ch02.en.html#_debian_archive_basics

The reason we set a default mirror for debootstrap is mostly historical as some versions would misbehave without it (iirc invalid apt sources list). That's something we could probably drop now so we just use the defaults for mmdebstrap/debootstrap.

> Pretty much all Debian tools had defaulted to http in the past, it's the only effective way to get caching: with https, apt has to establish an end to end encryption to a hostname with that cert's subject name, so either one has some use some tricks like http://HTTPS/// (apt-cacher-ng) or setup a root CA to do SSL interception.

Ack; it's also why debos support propagate proxy environment settings as some use that for caching.
 
> Yes, the debian archive is securely delivered over http as the indices are GPG signed and there is a timestamp (valid until) to prevent replay attacks.
> 
> What was brought up in the security assessment that trigger my proposed changes is a lack of confidentiality: I can observe the software being installed on this or that system. It's minor, but valid, and http traffic will generally trigger alerts.
> 
> > when building for non-native architectures using https can also cause image builds to be quite a bit slower as the whole encryption is emulated
> 
> Yeah, SSL at a large scale is costly for client and server (I guess we don't care too much about Fastly), good point about QEMU.
> 
> I guess it's a classical example of security vs practicality: http is generally nice for performance and caching, but will trigger security reviews and leak some data about systems. :-/

Agreed; This is one of those fuzzy areas and i can see both sides of. Though I'd prefer Debian to take the lead here and debos following the defaults.

> One thing we discussed with Robie for qcom-deb-images is to have an easy flag to override the mirror at build time, but keep https in the target images.

That makes total sense!. For this PR, we can merge the non-debootstrap commits. Is not setting a default debootstrap mirror something you'd like to look at?

obbardc commented on actions/debootstrap_action.go:
```diff
@@ -27,7 +27,7 @@ Optional properties:
 
   - mirror -- URL with Debian-compatible repository
-    If no mirror is specified debos will use http://deb.debian.org/debian as default.
+    If no mirror is specified debos will use https://deb.debian.org/debian as default.
```
```suggestion
    If no mirror is specified the default mirror of debootstrap will be used
```

obbardc reviewed (commented):
following Sjoerd's comments

lool committed:
tests: Add integration tests for debootstrap

Test with default components and mirror, and test overring default
mirror and components.

Signed-off-by: Loïc Minier <loic.minier@oss.qualcomm.com>

lool committed:
refactor(debootstrap_action): Add unit tests

Split debootstrap command generation in a separate function to add unit
tests.

Signed-off-by: Loïc Minier <loic.minier@oss.qualcomm.com>

lool committed:
refactor(debootstrap_action): Drop APT sources gen

debootstrap already generates sources.list and this requires hardcoding
the default mirror in debos, so just leave it to debootstrap.

Signed-off-by: Loïc Minier <loic.minier@oss.qualcomm.com>

lool committed:
refactor(debootstrap_action): No default mirror

Rely on debootstrap's default mirror instead of adding our own. Matches
mmdebstrap's behavior.

Signed-off-by: Loïc Minier <loic.minier@oss.qualcomm.com>

lool commented:
> @lool small bump

Thanks for the ping @obbardc! Here's an updated set of commits; perhaps the unit and integration tests are a bit overkill.

I've ran debootstrap manually to verify proper behavior, but while I started running the integration tests locally, they take a very long time with qemu (without kvm). I'm hoping the CI will do its magic :)

lool commented on actions/debootstrap_action.go:
```diff
@@ -86,7 +86,7 @@ func NewDebootstrapAction() *DebootstrapAction {
 	d.Components = []string{"main"}
 	// Set generic default mirror
-	d.Mirror = "http://deb.debian.org/debian"
+	d.Mirror = "https://deb.debian.org/debian"
```
yup; gave this some love

lool commented:
> That makes total sense!. For this PR, we can merge the non-debootstrap commits. Is not setting a default debootstrap mirror something you'd like to look at?

Did this in the updated PR.

TBH, I kind of feel that debos is not consistent between setting a default component (main) but not setting a default mirror, but with the proposed changes at least mmdebstrap and debootstrap actions have similar expectations and we're not hardcoding the debootstrap default mirror.

I was tempted to remove the default main component (also debootstrap's default) and even the "unstable" debootstrap script (the default for all supported suites).
